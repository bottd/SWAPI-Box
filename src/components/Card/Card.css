.Card {
  border: 6px solid black;
  border-radius: 20px;
  height: 260px;
  margin: 10px;
  overflow: scroll;
  padding: 10px;
  text-align: center;
  width: 280px;
}

ul {
  font-size: 18px;
  line-height: 2.2;
  list-style: none;
  padding-left: 5px;
}

.name {
  font-size: 24px;
  margin: 10px 5px;
  padding-bottom: 10px;
  border-bottom: 4px solid black;
}

.favoriteIcon {
  border: 4px solid black;
  border-radius: 20px;
  color: black;
  cursor: pointer;
  font-family: starWars;
  float: right;
  padding: 0 18px;
  position: relative;
  bottom: 14px;
}

.favoriteIcon:hover {
  background-color: black;
  border: 4px solid #fdde09;
  color: #fdde09;
  transition: all 0.4s ease-in-out;
}

.favorited {
  background-color: black;
  border: 4px solid #fdde09;
  color:#fdde09;
}
