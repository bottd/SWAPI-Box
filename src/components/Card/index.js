import React, { Component } from "react";
import PropTypes from "prop-types";
import "./Card.css";

class Card extends Component {
  constructor(props) {
    super(props);
    this.state = { favorited: this.props.cardData.favorited };
  }

  handleClick = () => {
    const { favorited } = this.state;
    this.props.cardData.favorited = !favorited;
    this.setState(
      { favorited: !favorited },
      this.props.handleCardClick(this.props.cardData, this.state.favorited)
    );
  };

  render() {
    const { favorited } = this.state;
    const { handleCardClick, cardData } = this.props;
    const { name } = cardData;
    const listItems = Object.keys(cardData).map((value, index) => {
      if (value === "name" || value === "favorited") return;
      if (cardData[value].length) {
        return <li key={name + index}>{`${value}: ${cardData[value]}`}</li>;
      }
    });
    return (
      <div className="Card">
        <h2 className="name">
          {name}
          <span
            className={"favoriteIcon" + (favorited ? " favorited" : "")}
            onClick={this.handleClick}
          >
            {favorited ? "#" : "$"}
          </span>
        </h2>
        <ul>{listItems}</ul>
      </div>
    );
  }
}

Card.propTypes = {
  handleCardClick: PropTypes.func.isRequired
};

export default Card;
