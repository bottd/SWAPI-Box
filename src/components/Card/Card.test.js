import React from "react";
import { shallow } from "enzyme";
import Card from "./index";

describe("Card", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Card
        handleCardClick={jest.fn()}
        cardData={{
          name: "Yoda",
          species: "Yoda's Species",
          favorited: false
        }}
      />
    );
  });

  it("Renders like snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Favorited should toggle when handleClick is called", async () => {
    await wrapper.instance().handleClick();
    expect(wrapper.state().favorited).toBe(true);
  });

  it("Favorited should toggle when handleClick called twice", async () => {
    await wrapper.instance().handleClick();
    await wrapper.instance().handleClick();
    expect(wrapper.state().favorited).toBe(false);
  });

  it("Should trim long names down", () => {
    wrapper = shallow(
      <Card
        handleCardClick={jest.fn()}
        cardData={{
          name: "Yoda except a little longer",
          favorited: false
        }}
      />
    );
    const name = wrapper.find(".name");
    expect(name.text()).toBe("Yoda except a little...$");
  });

  it("Should render extra keys in cardData as li elements", () => {
    const li = wrapper.find("li");
    expect(li.text()).toBe("species: Yoda's Species");
  });
});
