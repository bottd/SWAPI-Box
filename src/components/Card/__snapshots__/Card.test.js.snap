// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Renders like snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    cardData={
      Object {
        "name": "Yoda",
      }
    }
    handleCardClick={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="name"
        >
          Yoda
          <span
            className="favoriteIcon"
            onClick={[Function]}
          >
            $
          </span>
        </h2>,
        <ul />,
      ],
      "className": "Card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Yoda",
            <span
              className="favoriteIcon"
              onClick={[Function]}
            >
              $
            </span>,
          ],
          "className": "name",
        },
        "ref": null,
        "rendered": Array [
          "Yoda",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "$",
              "className": "favoriteIcon",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "$",
            "type": "span",
          },
        ],
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          undefined,
        ],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="name"
          >
            Yoda
            <span
              className="favoriteIcon"
              onClick={[Function]}
            >
              $
            </span>
          </h2>,
          <ul />,
        ],
        "className": "Card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Yoda",
              <span
                className="favoriteIcon"
                onClick={[Function]}
              >
                $
              </span>,
            ],
            "className": "name",
          },
          "ref": null,
          "rendered": Array [
            "Yoda",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "$",
                "className": "favoriteIcon",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "$",
              "type": "span",
            },
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            undefined,
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
